cpp/src/binary/jit/code/traits.cpp

```diff
@@ -0,0 +1,237 @@
+    struct HelperIntegralMap<1, IntegralSigned> {
+        using Type = int8_t;
+    };
+
+    template <>
+    struct HelperIntegralMap<2, IntegralSigned> {
+        using Type = int16_t;
+    };
+
+    template <>
+    struct HelperIntegralMap<4, IntegralSigned> {
+        using Type = int32_t;
+    };
+
+    template <>
+    struct HelperIntegralMap<8, IntegralSigned> {
+        using Type = int64_t;
+    };
+
+    template <>
+    struct HelperIntegralMap<1, IntegralUnsigned> {
+        using Type = uint8_t;
+    };
+
+    template <>
+    struct HelperIntegralMap<2, IntegralUnsigned> {
+        using Type = uint16_t;
+    };
+
+    template <>
+    struct HelperIntegralMap<4, IntegralUnsigned> {
+        using Type = uint32_t;
+    };
+
+    template <>
+    struct HelperIntegralMap<8, IntegralUnsigned> {
+        using Type = uint64_t;
+    };
+
+    template <int N, typename T>
+    using IntegralMap = typename HelperIntegralMap<N, T>::Type;
+
+
+
+    template <bool, typename, typename>
+    struct helperIf;
+
+    template <typename T, typename F>
+    struct helperIf<true, T, F> {
+        using type = T;
+    };
+
+    template <typename T, typename F>
+    struct helperIf<false, T, F> {
+        using type = F;
+    };
+
+    template <bool B, typename T, typename F>
+    using If = typename helperIf<B, T, F>::type;
+
+
+
+    template<bool B, class T>
+    struct helperEnableIf
+    {};
+
+    template<class T>
+    struct helperEnableIf<true, T> {
+        using type = T;
+    };
+
+    template <bool B, class T = void>
+    using enableIf = typename helperEnableIf<B,T>::type;
+
+
+
+    template <typename Vax, typename Vay, typename = void>
+    struct HelperCommonNumber {};
+
+    template <typename Vax, typename Vay>
+    struct HelperCommonNumber<Vax, Vay, enableIf<(isFloatingPoint<Vax> || isFloatingPoint<Vay>)>> {
+        using Type = If<(sizeof(Vax) == 8 || sizeof(Vay) == 8), double, float>;
+    };
+
+    template <typename Vax, typename Vay>
+    struct HelperCommonNumber<Vax, Vay, enableIf<(isIntegralSigned<Vax> && isIntegralSigned<Vay>)>> {
+        using Type = IntegralMap<(MaxSize<Vax, Vay>), IntegralSigned>;
+    };
+
+    template <typename Vax, typename Vay>
+    struct HelperCommonNumber<Vax, Vay, enableIf<(isIntegralSigned<Vax> && isIntegralUnsigned<Vay>)>> {
+        using Type = IntegralMap<(MaxSize<Vax, Vay>), IntegralSigned>;
+    };
+
+    template <typename Vax, typename Vay>
+    struct HelperCommonNumber<Vax, Vay, enableIf<(isIntegralUnsigned<Vax> && isIntegralSigned<Vay>)>> {
+        using Type = IntegralMap<(MaxSize<Vax, Vay>), IntegralSigned>;
+    };
+
+    template <typename Vax, typename Vay>
+    struct HelperCommonNumber<Vax, Vay, enableIf<(isIntegralUnsigned<Vax> && isIntegralUnsigned<Vay>)>> {
+        using Type = IntegralMap<(MaxSize<Vax, Vay>), IntegralUnsigned>;
+    };
+
+    template <typename Vax, typename Vay>
+    using CommonNumber = typename HelperCommonNumber<Vax, Vay>::Type;
```

Comment:
All of this custom `CommonNumber` logic can be replaced with `std::common_type`.

It doesn't look like Jitify has this trait yet, but it should be added as it would vastly simplify this code.

https://en.cppreference.com/w/cpp/types/common_type

Reply:
Solved by removing `CommonNumber` altogether.